### ISSUE TITLE
Add health-check endpoints

### ISSUE THREAD

**Issue by bzz** [src-d/lookout#180]
Server and Analyzer should expose an HTTP-level endpoints that can be used as kubernetes life-probes.

As [discussed](https://github.com/src-d/minutes/pull/309/files#diff-d707b85ec18906cb2d5bedb7373d595aR47), at this point probes could check accessibility of it's dependencies like DB, Bblfsh, etc.

Eventually, it would be really nice for such checks to become more complex and execute some basic scenarios and verify them pass (like a simple live unit-tests).

For now:
 - [ ] check what kind of life probes k8s supports/recomends
 - [ ] add endpoints to Server and Dummy analyzer
 - [ ] update charts \w life probes

**Comment by smacker** [comment]
@carlosms We started a discussion in 2 different places: PR and slack. I think it's better to move it here.

The discussions touch different topics, I'll try to summarize them here and give my opinion. I'll start from the "end", specific issues, and move to more general topics.

### Proper handling of restarts. Especially posting comments on github.

The main argument about it I noticed: restarting on the middle of results posting would cause problems.

Restart is a very common problem in software development. No matter do they caused intentionally or not. That's why almost all systems already have solutions for it. For databases it's transaction. For a queue it's publish confirmations. Those concepts allow to handle restarts in the middle of something be safe. Rollback would kick and an application can retry the same operation after the restart.

HTTP APIs are usually designed with this issue in mind also. Retry would cause specific error which can be easily handled by the application. Or API would provide some kind of transaction/confirmation also.

In our particular case "posting comments" github provides such API which allows retry without any undesired effects for an end user (like duplicated comments).
If we create [review](https://developer.github.com/v3/pulls/reviews/#create-a-pull-request-review) without `body` attribute, it will be visible to an user only after the [submit call](https://developer.github.com/v3/pulls/reviews/#submit-a-pull-request-review).

It's safe to call `submit` multiple times.

To make my point more clear that restart in the middle of posting isn't a problem here is a list of changes how we might improve our poster in the future:
- Call `create` without body if don't have ID of review in the state already
- Save returned ID to the state
- In case of failure, we can just retry later without any issues for an end user
- Send confirm
- Update status in DB
- In case of failure, we can just safely retry later

### Should we consider DB as a critical component of the system or not

The current argument:
> Maybe a restart is the better option from our point of view, but from management it's uptime... maybe we prefer to keep posting analysis while the DB is down, and deal with the inconsistent state somehow later when the DB comes up again

According to our requirements, we can't post comments without the state be available. It would cause duplicated comments. Our current codebase doesn't allow not only to post comments but event watch without DB at all. As long as we must satisfy non-duplicated comments business requirement we have to have some state available. For now, it's DB. So I propose to treat it is a critical component from the start.

### Should we consider Babelfish as a critical component of the system or not

Our current production analyzers require bblfsh. During the retreat, all but 1 project required bblfsh which makes me think in terms of the whole product that bblfsh is the main component.

Another reason to treat it as a critical component is to avoid the domino effect. Unavailable bblfsh would cause errors not only on our side but on analyzers side too.
Though we can avoid it by introducing more complexity like remembering which analyzers require bblfsh and send requests only to ones without this requirement, adding [CircuitBreakers](https://martinfowler.com/bliki/CircuitBreaker.html), retries and other stuff. But restarts also solve these problems and much easier to implement on the current stage of the project.

For now I would fail in case bblfsh isn't available and improve it later with better solutions.

### The idea of complete graceful handling errors vs let it crash

Nobody argues that errors which can be handled gracefully should be handled like that.
Though there are lots of errors which we as developers can't think about during development. And it gets much worse as soon as the software includes more than 1 component and even worse when we make it distributed.

Apparently, most of the bugs which happen in production are [heisenbugs](https://en.wikipedia.org/wiki/Heisenbug). In the [Why Do Computers Stop and What Can Be Done About It?](http://www.hpl.hp.com/techreports/tandem/TR-85.7.pdf) paper author discovered in his experiments that only 1 bug out of 132 was caused by bug which can be easily reproduced and restart doesn't fix it.

Kubernetes gives us power of fast automatic restart and allow to handle critical errors in a similar way to Erlang. Supervisor would restart services without human involvement which solves a lot of problems. Here an example of a known bug and a theoretical bug which it solves but there are tons of other bugs we didn't discover yet or can't imagine.

- we clone repositories and put them on disk. But never clean. Sooner or later container would run out of free space. The restart solves it without waking up Rafa or us in the middle of the night.
- we could forget to close a transaction in some cases. This would lead to running out of available connections. The restart solves it without waking up Rafa or us in the middle of the night.

Supervisor with restarts increases robustness, availability and reliability.
This is the reason why this issue exists in the first place.

---

I also add few links here you might be interested to check:
- [Pattern: Health Check API](https://microservices.io/patterns/observability/health-check-api.html)
- [Addressing Cascading Failures](https://landing.google.com/sre/book/chapters/addressing-cascading-failures.html)
- [Kubernetes best practices: Setting up health checks with readiness and liveness probes](https://cloudplatform.googleblog.com/2018/05/Kubernetes-best-practices-Setting-up-health-checks-with-readiness-and-liveness-probes.html)
- [What Erlang Taught Me About Distributed Systems](https://blog.kenforthewin.com/what-erlang-taught-me-about-distributed-systems/)
- [The Zen of Erlang](https://ferd.ca/the-zen-of-erlang.html)

**Comment by dpordomingo** [comment]
There is already (for workers and watcher):
- `/health/liveness`, returns `200` if it was started
- `/health/readiness`, returns `200` if all startup instructions succeded

And for `dummy`:
- `/health/liveness`, returns `200` if it was started
- `/health/readiness`, returns `200` if it was started

Is it :point_up: enough for second subtask?
>_ add endpoints to Server and Dummy analyzer_

**Comment by smacker** [comment]
I still think DB & Bblfsh are critical components for lookout and we should report non 200 status when they are not available.

Plus the concern of a restart in the middle of posting was solved already by syncing comments with github in case we realized the previous event wasn't finished correctly.

**Comment by dpordomingo** [comment]
I agree with you @smacker about criticalness of DB and bblfsh.
would you create two new endpoints for them? or should we adapt /health/readiness to also check the DB and bblfsh?

**Comment by smacker** [comment]
I don't think you can [define](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-a-liveness-http-request) more than 1 endpoint for readness and liveness checks in kubernetes. So it should be the same endpoints.

**Comment by rporres** [comment]
If there's any issue with any app hard requirement it makes sense that it does not receive requests. The main issue is to determine if the application it should be restarted for that or not, as a readiness probe fail in k8s won't force the pod restart.
